### DISCUSSION TITLE
Why doesn't `copilot#AcceptWord` and `copilot#AcceptLine` work when called from Lua?

### DISCUSSION THREAD

**Discussion by Hubro** [community/community#112974] [Copilot Conversations] [answered]
### Select Topic Area

Question

### Body

Using `<Plug>(copilot-accept-word/line)` works when I remap to them, but as a typical (n)vim addict, I want to customize my experience by adding more logic before/after accepting the word/line.

Since (AFAIK) I can't directly invoke `<Plug>(copilot-accept-*)` without mapping a key to it, I instead checked the source code and found that all the mapping does is call `copilot#AcceptWord/Line()`.

With that in mind, I tried making my own mapping in Lua:

```lua
local accept_word = function()
  -- TODO: pre-stuff
  vim.fn["copilot#AcceptWord"]()
  -- TODO: post-stuff
end

vim.keymap.set("i", "<A-l>", accept_word, { remap = true, replace_keycodes = false })
```

Immediate failure. All `vim.fn["copilot#AcceptWord"]()` does is dismiss the suggestion. No text is inserted.

How come these functions work when invoked through `<Plug>(copilot-...)` but don't work when invoked directly?

---

Just for context, the primary issue I have with the basic `<Plug>(copilot-accept-*)` mapping is that they dismiss the suggestion. I almost always want to accept *more* than 1 word or line of the suggestion, so it's a waste of time and key strokes to have to invoke `copilot#Suggest` after every word. (Interestingly, `copilot#Suggest` *does* work when called directly.)

---

**EDIT:** I spent waaay too much time on this, but I finally figured out a working workaround:

```lua
local _plug_accept = function(thing)
  local expr = vim.api.nvim_replace_termcodes(
    "<Plug>(copilot-accept-" .. thing .. ")", true, true, true
  )
                                                               
  vim.api.nvim_feedkeys(expr, "m", false)
end
                                                               
local suggest = function()
  require("cmp").abort() -- Hide cmp completion window
  vim.fn["copilot#Suggest"]()
end
                                                               
local accept_word = function()
  _plug_accept("word")
                                                               
  vim.schedule(suggest)
end
                                                               
local accept_line = function()
  _plug_accept("line")
                                                               
  vim.schedule(suggest)
end

vim.keymap.set("i", "<A-p>", suggest, { silent = true })
vim.keymap.set("i", "<C-e>", 'copilot#Accept("")', { expr = true, replace_keycodes = false })
vim.keymap.set("i", "<A-l>", accept_word)
vim.keymap.set("i", "<A-h>", accept_line)
```

It's ugly and suboptimal, but it works.

**Comment by tpope** [comment, marked as answer]
> I instead checked the source code and found that all the mapping does is call `copilot#AcceptWord/Line()`.

It doesn't just call it, it returns the value, which ends up in the `<expr>` map. You need to do the same.

**Reply by Hubro** [reply]
Oh! I had absolutely *no* idea the callback could return a value, that explains a lot. Thanks!
